File: services/graphql-crm/src/mutation/leads.rs
use async_graphql::{Context, InputObject, Object};
use chrono::DateTime;
use chrono::Utc;
use uuid::Uuid;

use crate::models::enums::{LeadSource, LeadStatus};
use crate::models::leads;

#[derive(Debug, Clone, InputObject)]
pub struct CreateLeadInput {
    pub name: String,
    pub email: String,
    pub lead_source: Option<LeadSource>,
    pub status: Option<LeadStatus>,
    pub lead_score: Option<i32>,
    pub owner_id: Uuid,
    pub campaign_id: Option<Uuid>,
    pub converted_at: Option<DateTime<Utc>>,
    pub converted_contact_id: Option<Uuid>,
    pub converted_company_id: Option<Uuid>,
    pub converted_opportunity_id: Option<Uuid>,
}

#[derive(Debug, Clone)]
pub struct Mutation;

#[Object(name = "CrmLeadsMutations")]
impl Mutation {
    async fn create_lead(
        &self,
        ctx: &Context<'_>,
        payload: CreateLeadInput,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_name(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        name: String,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_email(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        email: String,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_lead_source(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        lead_source: Option<LeadSource>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_status(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        status: Option<LeadStatus>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_lead_score(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        lead_score: Option<i32>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_owner_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        owner_id: Uuid,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_campaign_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        campaign_id: Option<Uuid>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_converted_at(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        converted_at: Option<DateTime<Utc>>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_converted_contact_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        converted_contact_id: Option<Uuid>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_converted_company_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        converted_company_id: Option<Uuid>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn update_lead_converted_opportunity_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        converted_opportunity_id: Option<Uuid>,
    ) -> async_graphql::Result<leads::Model> {
        todo!()
    }
    async fn remove_lead(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        todo!()
    }
}

File: src/graphql/mutations/crm/leads.ts
import { graphql } from "@/lib/graphql/client";

export const createLead = graphql(`
  mutation CreateLead($payload: CreateLeadInput!) {
    crm {
      createLead(payload: $payload) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadName = graphql(`
  mutation UpdateLeadName($id: UUID!, $name: String!) {
    crm {
      updateLeadName(id: $id, name: $name) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadEmail = graphql(`
  mutation UpdateLeadEmail($id: UUID!, $email: String!) {
    crm {
      updateLeadEmail(id: $id, email: $email) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadLeadSource = graphql(`
  mutation UpdateLeadLeadSource($id: UUID!, $leadSource: LeadSource) {
    crm {
      updateLeadLeadSource(id: $id, leadSource: $leadSource) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadStatus = graphql(`
  mutation UpdateLeadStatus($id: UUID!, $status: LeadStatus) {
    crm {
      updateLeadStatus(id: $id, status: $status) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadLeadScore = graphql(`
  mutation UpdateLeadLeadScore($id: UUID!, $leadScore: Int) {
    crm {
      updateLeadLeadScore(id: $id, leadScore: $leadScore) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadOwnerId = graphql(`
  mutation UpdateLeadOwnerId($id: UUID!, $ownerId: UUID!) {
    crm {
      updateLeadOwnerId(id: $id, ownerId: $ownerId) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadCampaignId = graphql(`
  mutation UpdateLeadCampaignId($id: UUID!, $campaignId: UUID) {
    crm {
      updateLeadCampaignId(id: $id, campaignId: $campaignId) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadConvertedAt = graphql(`
  mutation UpdateLeadConvertedAt($id: UUID!, $convertedAt: DateTime) {
    crm {
      updateLeadConvertedAt(id: $id, convertedAt: $convertedAt) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadConvertedContactId = graphql(`
  mutation UpdateLeadConvertedContactId($id: UUID!, $convertedContactId: UUID) {
    crm {
      updateLeadConvertedContactId(id: $id, convertedContactId: $convertedContactId) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadConvertedCompanyId = graphql(`
  mutation UpdateLeadConvertedCompanyId($id: UUID!, $convertedCompanyId: UUID) {
    crm {
      updateLeadConvertedCompanyId(id: $id, convertedCompanyId: $convertedCompanyId) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const updateLeadConvertedOpportunityId = graphql(`
  mutation UpdateLeadConvertedOpportunityId($id: UUID!, $convertedOpportunityId: UUID) {
    crm {
      updateLeadConvertedOpportunityId(id: $id, convertedOpportunityId: $convertedOpportunityId) {
        id
        name
        email
        leadScore
        leadSource
        status
        convertedAt
        createdAt
        updatedAt
        owner {
          id
          name
        }
        campaign {
          id
          name
        }
        convertedContact {
          id
          name
        }
        convertedOpportunity {
          id
          name
        }
      }
    }
  }
`);

export const removeLead = graphql(`
  mutation RemoveLead($id: UUID!) {
    crm {
      removeLead(id: $id)
    }
  }
`);
